src/libsystemd/sd-device/device-internal.h

```diff
@@ -47,6 +47,10 @@ struct sd_device {
         uint64_t devlinks_iterator_generation; /* generation when iteration was started */
         int devlink_priority;
 
+        Hashmap *children;
+        Iterator children_iterator;
+        bool children_enumerated;
```

Comment:
Hmmm, let's talk about the big picture. I know that this follows a pattern that was used before, but putting the iterator in the structure blows up the struct size, and allows only one iteration at a time. Shouldn't we instead return an opaque `sd_iterator` object from `sd_device_get_child_first()`, and pass it into `sd_device_get_child_next()` ?

Reply:
if memory usage really is a problem I guess we could split out parts of the sd_device structure into separate structures, and then allocate it only on first use. But not sure it's worth doing that wihtout some memory profiling first...

And i guess the devlinks and tags iterators might be better candidates for that, given they have much more fields...